### PR TITLE
Spike: returning a record replacement with glide strings.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
TODO:
- [ ] Complete code todos
- [ ] Complete updating commands:
  - [x] xinfo streams
  - [ ] function stats
  - [ ] function list
  - [ ] commands which return `ClusterValue`
- [x] Update examples in docs
- [x] Update description of returned types where changed
- [ ] Update examples - indirect changes (especially `zadd` - used in examples of other commands)
- [x] Update transaction
- [ ] Fix tests
- [x] Update exports
- [ ] Add new tests
- [ ] Update doc for custom command

List of commands:
* `LMPOP`
* `BLMPOP`
* `BZMPOP`
* `ZPOPMIN`
* `ZPOPMAX`
* `ZMPOP`
* `ZDIFFWITHSCORES`
* `ZINTERWITHSCORES`
* `ZUNIONWITHSCORES`
* `HGETALL`
* `LSCIDX` - no return type change
* `PUBSUB NUMSUB`
* `XRANGE` - no return type change
* `XREVRANGE` - no return type change
* `XINFO GROUPS` - no return type change
* `XINFO CONSUMERS` - no return type change
* `XINFO STREAM` - no return type change
* `XCLAIM` - no return type change
* `XAUTOCLAIM` - no return type change
* `XREAD`
* `XREADGROUP`

### CODE REVIEW

File: node/rust-client/src/lib.rs [hunk 1]
```diff
@@ -195,13 +195,18 @@ fn redis_value_to_js(val: Value, js_env: Env, string_decoder: bool) -> Result<Js
...
         }
         Value::Map(map) => {
-            let mut obj = js_env.create_object()?;
-            for (key, value) in map {
+            // Convert map to array of key-value pairs instead of a `Record`` (object),
+            // because `Record` does not support `GlideString` as a key.
+            let mut js_array = js_env.create_array_with_length(map.len())?;
+            for (idx, (key, value)) in (0_u32..).zip(map.into_iter()) {
+                let mut obj = js_env.create_object()?;
                 let field_name = String::from_owned_redis_value(key).map_err(to_js_error)?;
```
Comment (acarbonetto) [inline node/rust-client/src/lib.rs:203]: I think you need to apply decoder: 
```
            if string_decoder {
                let str = to_js_result(std::str::from_utf8(data.as_ref()))?;
                field_name = js_env.create_string(str).map(|val| val.into_unknown())?;
            } else {
                field_name = js_env.create_buffer_with_data(data)?.into_unknown());
            }
```

File: node/tsconfig.json [hunk 2]
```diff
@@ -25,6 +25,15 @@
...
         "outDir": "./build-ts" /* Specify an output folder for all emitted files.*/
     },
+    "ts-node": {
+        "transpileOnly": true,
+        "compilerOptions": {
+            "module": "CommonJS",
+            "target": "ES2018",
+            "esModuleInterop": true
+        },
+        "esm": true
```
Comment (acarbonetto) [inline node/tsconfig.json:35]: Does Interloping have any detrimental performance implications?  
I'd be concerned about adding this to the project.  It's kind of a development feature.

Comment (Yury-Fridlyand) [reply node/tsconfig.json:35]: Will revert this file, added for testing

File: node/src/BaseClient.ts [hunk 3]
```diff
@@ -1781,23 +1814,37 @@ export class BaseClient {
...
      */
-    public async hgetall(key: string): Promise<Record<string, string>> {
-        return this.createWritePromise(createHGetAll(key));
+    public async hgetall(
+        key: string,
+        options?: DecoderOption,
+    ): Promise<{ field: GlideString; value: GlideString }[]> {
+        return this.createWritePromise(createHGetAll(key), options).then(
+            (res) =>
+                (res as GlideRecord<GlideString>).map((r) => {
```
Comment (Yury-Fridlyand) [inline node/src/BaseClient.ts:1844]:
```suggestion
        return this.createWritePromise<GlideRecord<GlideString>>(createHGetAll(key), options).then(
            res =>
                res.map((r) => {
```

File: node/src/BaseClient.ts [hunk 4]
```diff
@@ -3351,9 +3400,19 @@ export class BaseClient {
...
-        return this.createWritePromise(createXRange(key, start, end, count));
+        options?: { count?: number } & DecoderOption,
+    ): Promise<{ entryID: GlideString; data: [string, string][] }[] | null> {
+        return this.createWritePromise(
+            createXRange(key, start, end, options?.count),
+            options,
+        )
+            .then((res) => res as GlideRecord<[string, string][]> | null)
+            .then(
+                (res) =>
```
Comment (Yury-Fridlyand) [inline node/src/BaseClient.ts:3411]:
```suggestion
        return this.createWritePromise<GlideRecord<[string, string][]> | null>(
            createXRange(key, start, end, options?.count),
            options,
        )
            .then(
                res =>
```

File: node/src/BaseClient.ts [hunk 5]
```diff
@@ -3391,48 +3452,62 @@ export class BaseClient {
         key: string,
         end: Boundary<string>,
         start: Boundary<string>,
-        count?: number,
-    ): Promise<Record<string, [string, string][]> | null> {
-        return this.createWritePromise(createXRevRange(key, end, start, count));
+        options?: { count?: number } & DecoderOption,
+    ): Promise<{ entryID: GlideString; data: [string, string][] }[] | null> {
```
Comment (Yury-Fridlyand) [inline node/src/BaseClient.ts:3456]: define a type

File: node/src/BaseClient.ts [hunk 6]
```diff
@@ -6584,8 +6812,15 @@ export class BaseClient {
      */
     public async pubsubNumSub(
         channels?: string[],
-    ): Promise<Record<string, number>> {
-        return this.createWritePromise(createPubSubNumSub(channels));
+        options?: DecoderOption,
```
Comment (Yury-Fridlyand) [inline node/src/BaseClient.ts:6815]: update doc

File: node/src/BaseClient.ts [hunk 7]
```diff
@@ -297,6 +297,35 @@ export type DecoderOption = {
     decoder?: Decoder;
 };
 
+/** A replacement for `Record<GlideString, T>` - array of key-value pairs. */
+export type GlideRecord<T> = { key: GlideString; value: T }[];
+
+// GLideRecord for score for sorted set commands?
+// TODO export
+export type ElementScoreData = { element: GlideString; score: number }[];
```
Comment (acarbonetto) [inline node/src/BaseClient.ts:305]: Suggest to call it SortedSetDataType or something similar.  
For Stream, we can call it StreamEntryDataType?

Comment (adarovadya) [inline node/src/BaseClient.ts:305]: yes, I agree. This way you explicitly convey that this data type is meant for storing elements and their associated scores. This is more explicit than using GlideRecord<number> , where the meaning of "key" and "value" might be less obvious without additional comments

File: node/src/BaseClient.ts [hunk 8]
```diff
@@ -297,6 +297,35 @@ export type DecoderOption = {
...
+// TODO remove unknown, use <>
+function convertGlideRecordForZSet(
+    res: unknown | GlideRecord<number>,
+): ElementScoreData {
+    return (res as GlideRecord<number>).map((e) => {
+        return { element: e.key, score: e.value };
+    });
+}
+
+/** Downcast `GlideRecord` to `Record`. Use if you are 146% aware that `data` keys are always strings. */
```
Comment (adarovadya) [inline node/src/BaseClient.ts:320]: 146?

File: node/src/BaseClient.ts (hunk 5 above)
Comment (adarovadya) [inline node/src/BaseClient.ts:3456]: Does the data not need to be a GlideString?

File: node/src/BaseClient.ts [hunk 9]
```diff
@@ -297,6 +297,35 @@ export type DecoderOption = {
...
+// GLideRecord for score for sorted set commands?
+// TODO export
+export type ElementScoreData = { element: GlideString; score: number }[];
+
+/**
+ * Convert `GlideRecord<number>` recevied after resolving the value pointer into `ElementScoreData`.
+ * `createWritePromise` returns `unknown`, so we need extra cast and caution.
+ */
+// TODO remove unknown, use <>
+function convertGlideRecordForZSet(
```
Comment (acarbonetto) [inline node/src/BaseClient.ts:312]: let's move this another PR - and just do some simple examples here.

File: node/src/BaseClient.ts [hunk 10]
```diff
@@ -1210,7 +1239,8 @@ export class BaseClient {
         );
     }
 
-    /** Retrieve the values of multiple keys.
+    /**
+     * Retrieves the values of multiple keys.
```
Comment (acarbonetto) [inline node/src/BaseClient.ts:1243]: would be really good to move all these peripheral changes to another PR
so we can focus on what's important:  the Record refactor

File: node/src/BaseClient.ts (hunk 5 above)
Comment (Yury-Fridlyand) [reply node/src/BaseClient.ts:3456]: Yes, it is a user data

File: node/src/BaseClient.ts [hunk 11]
```diff
@@ -297,6 +297,35 @@ export type DecoderOption = {
     decoder?: Decoder;
 };
 
+/** A replacement for `Record<GlideString, T>` - array of key-value pairs. */
+export type GlideRecord<T> = { key: GlideString; value: T }[];
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply node/src/BaseClient.ts:301]: Why? Without `[]` it stores only one K-V pair

File: node/tsconfig.json [hunk 12]
```diff
@@ -25,6 +25,15 @@
...
+    },
```
Comment (Yury-Fridlyand) [inline node/tsconfig.json:36]:
```suggestion
```

File: node/src/BaseClient.ts (hunk 8 above)
Comment (Yury-Fridlyand) [reply node/src/BaseClient.ts:320]: When 100% is not enough

File: node/src/BaseClient.ts (hunk 5 above)
Comment (Yury-Fridlyand) [reply node/src/BaseClient.ts]: Update: entry ID could be given by the user, but it is still in number-like format (to be precise: `"<uint>-<uint>"`). I updated it by introducing
```ts
export type StreamEntryDataType = Record<string, [GlideString, GlideString][]>;
```
Data is user data, which could contain anything

File: node/src/BaseClient.ts (hunk 10 above)
Comment (Yury-Fridlyand) [reply node/src/BaseClient.ts:1270]: Reverted
